shared/docs/PROTOCOL_STATE_HASH.md

```diff
@@ -0,0 +1,173 @@
+
+On every round of every height:
+
+1. The `Consensus` module handles a `NEWROUND` message
+2. A new `UtilityContext` is initialized at the current height
+3. A new `PersistenceRWContext` is initialized at the current height
+4. The [Block Application](#block-application) flow commences
+
+```mermaid
+sequenceDiagram
+    title Steps 1-4
+    participant B as Bus
+    participant C as Consensus
+    participant U as Utility
+    participant P as Persistence
+
+    %% Handle New Message
+    B-->>C: HandleMessage(msg)
+
+    critical NewRound
+        %% Create Contexts
+        C->>+U: NewContext(height)
+        U->>+P: NewRWContext(height)
+        P->>-U: PersistenceContext
+        U->>U: store context<br>locally
+        activate U
+        deactivate U
+        U->>-C: UtilityContext
+        C->>C: store context<br>locally
+        activate C
+        deactivate C
+
+        %% Apply Block
+        Note over C, P: See 'Block Application'
+    end
+```
+
+5. The **HotPOKT lifecycle** takes place so Validators achieve consensus (i.e. steps `PRECOMMIT` and `COMMIT`)
+6. The `Consensus` module handle the `DECIDE` message
+7. The final `quorumCertificate` is propagated to the `UtilityContext` on Commit
+8. The final `quorumCertificate` is propagated to the `PersistenceContext` on Commit
+9. The persistence module's internal implementation for [Store Block](../../persistence/docs/PROTOCOL_STATE_HASH.md) must execute.
+10. Both the `UtilityContext` and `PersistenceContext` are released
+
+```mermaid
+sequenceDiagram
+    title Steps 6-10
+    participant B as Bus
+    participant C as Consensus
+    participant U as Utility
+    participant P as Persistence
+
+    %% Handle New Message
+    B-->>C: HandleMessage(msg)
+
+    critical Decide
+        %% Commit Context
+        C->>+U: Context.Commit(quorumCert)
+        U->>+P: Context.Commit(quorumCert)
+        P->>P: Internal Implementation
+        Note over P: Store Block
+        P->>-U: err_code
+        U->>C: err_code
+        deactivate U
+
+        %% Release Context
+        C->>+U: Context.Release()
+        U->>+P: Context.Release()
+        P->>-U: err_code
+        U->>-C: err_code
+    end
+
+```
+
+## Block Application
+
+When applying the block during the `NEWROUND` message shown above, the majority of the flow is similar between the _leader_ and the _replica_ with one of the major differences being a call to the `Utility` module as seen below.
+
+- `ApplyBlock` - Uses the existing set of transactions to validate & propose
+- `CreateAndApplyProposalBlock` - Reaps the mempool for a new set of transaction to validate and propose
+
+```mermaid
+graph TD
+    B[Should I prepare a new block?] --> |Wait for 2/3+ NEWROUND messages| C
+
+    C[Am I the leader?] --> |Yes| D
+    C[Am I the leader?] --> |No| Z
+
+    D[Did I get any prepareQCs?] --> |Find highest valid PrepareQC| E
```

Comment:
Question:

Does `HighPrepareQC` imply a `LockedBlock`?

Reply:
If a proposer dies before they submit the block it's a `locked_qc`. So a `locked_qc` can be a `high_prepare_qc`?

Reply:
If not, let's include it in your chart

Reply:
> So a `locked_qc` can be a `high_prepare_qc`

Yes.  A `lockedQC` may be either `nil` or `prepareQC`, and since we aggregate `prepareQCs`, it may be a `highPrepareQC`. It looks something like this:

| QC        | NewRound | Prepare | Precommit | Commit | Decide    |
| --------- | -------- | ------- | --------- | ------ | --------- |
| prepareQC | nil      | nil     | QC        | QC     | QC -> nil |
| lockedQC  | nil      | nil     | nil       | QC     | QC -> nil |

--- 

I updated the diagram to make it more explicit but not sure if answered the question directly. Lmk